### PR TITLE
Feat(audiobooks): Add support for OPUS audiobooks

### LINKED ISSUE #2908: [Feature] Opus support for audiobooks
### What's Your Idea?

Thank you for the audiobook support! This will be a joy to explore and utilize 🎧

My audiobooks library is somewhat "uncomplete" and unoptimized. Way back, I used Jellyfin for my audiobooks and decided to convert all my .mp3 files to opus. No chapters and no fancy audiobook formats - just plain audio files.

I would love to be able to play the opus converted files through BookLore. Any chance of implementing support for opus?

### Why Would This Be Helpful?

Support for audiobook libraries in formats other than M4B, M4A and .mp3.

### Anything Else? (Optional)

I have been using Storyteller on and off to merge epub files and opus files to create Readaloud versions of my books. Storyteller seems to accept opus as an audiobook format. If I remember correctly - they even have the option to convert existing .mp3 files to opus to create the Readaloud?

I would be happy to test the feature and report feedback while in development.

### Want to Help Out?

I'm happy to help test and give feedback

### Have You Considered Any Alternatives? (Optional)

I guess I could use another tool to batch convert my audiobook library to "native" audiobook formats, such as M4B, M4A and such. But since I'm not currently using any such software, It would be quite the process for me. But maybe the recommendations is to do just that, instead of implementing direct support for opus?

### Before Submitting

- [x] I've searched existing issues and confirmed this feature hasn't been requested yet

### PR DESCRIPTION
**Pull request by maxbengtzen**
## 📝 Description

Since my own audiobook library is made up of 100% .opus files, I thought that I would try and make Booklore support it. My guess (without having any statistics on it) is that I'm not the only one with this setup.

Including .opus in uploads and GUI was quite simple. It was just a matter of including the supported filetype associated with audiobooks and MIME mappings and such.

The audiobook player was a different matter. The current fork of jaudiotagger didn't support .opus like it support the currently supported file formats. I tried to mitigate this lack of support by tricking jaudiotagger that the opus files were .ogg files and it quickly became problematic. Lots of support files and scripts which didn't work. But then, I found another - more up to date - fork of jaudioagger - made by RouHim. This natively supports .opus (and Ogg Vorbis, FLAC, WMA, MP4, MPC, Monkey's Audio, OptimFROG, True Audio, WavPack, Musepack, Speex, WAV, AIFF, AU, RealAudio, Opus, and MP2 - which can now also be supported I guess). This change of fork made the PR quite easy even on the backend side of things.

**Linked Issue:** Fixes #2908 

> **Required.** Every PR must reference an approved issue. If no issue exists, [open one](https://github.com/booklore-app/booklore/issues/new) and wait for maintainer approval before submitting a PR. Unsolicited PRs without a linked issue will be closed.

## 🏷️ Type of Change

- [X] Bug fix
- [X] New feature
- [X] Enhancement to existing feature
- [ ] Refactor (no behavior change)
- [ ] Breaking change (existing functionality affected)
- [ ] Documentation update

## 🔧 Changes

- Replaced the current `jaudiotagger` dependency with the RouHim fork to enable Opus metadata parsing.
- Added `.opus` as a supported audiobook extension in backend file type/extension handling.
- Updated audiobook metadata extraction to route `.opus` through the same extractor flow as existing audiobook formats.
- Added `.opus` handling for audio file discovery, folder-based audiobook detection, fingerprinting, and first-audio-file resolution.
- Added `audio/opus` MIME type handling for streaming and OPDS/media responses.
- Updated frontend upload accept lists and file type classification so `.opus` files are treated as audiobooks.
- Updated frontend UI indicators/icons/styles so `.opus` appears consistently in book cards, metadata views, and related audiobook UI.
- Fixed audiobook player rendering so the page does not crash if audiobook info fails to load.
- Hardened `AudioMetadataService` against missing/nullable Opus audio header fields from `jaudiotagger` by treating bitrate/sample rate/channels/encoding as optional instead of failing the whole request.
- Added backend regression tests covering supported `.opus` extensions and missing-header-field scenarios.

## 🧪 Testing (MANDATORY)

> **PRs without this section filled out will be closed.** "Tests pass" or "Tested locally" is not sufficient. You must provide specifics.

**Manual testing steps you performed:**

1. Started the local dev stack with `dev.docker-compose.yml`.
2. Uploaded .opus file through Book Upload and Bookdrop via GUI and by adding the .opus file via putting the file in the Bookdrop folder.
3. Imported the book, adjusted metadata and made the audiobook "complete" with author and metadata that wasn't included in the file.
4. Watched the audiobook in the Library with the "opus" label applied.
5. Played the audiobook with the built in audiobook player
6. Tested downloading the .opus file from the book detail page
7. Monitored the backend logs while doing the above

**Regression testing:**

- Verified existing audiobook metadata/unit tests still pass after adding `.opus` support.
- Verified audio utility and OPDS MIME type tests still pass with the new extension handling.
- Confirmed the defensive header-field fallback does not prevent audiobook info generation when optional metadata is missing.

**Edge cases covered:**

- Single-file audiobook packaged as one `.opus` file.
- Opus files where `AudioHeader` fields like bitrate/sample rate/channels are missing or throw when accessed.
- UI behavior when `/api/v1/audiobook/{id}/info` fails, to avoid a secondary frontend crash.

**Test output:**

<details>
<summary>Backend test output (<code>./gradlew test</code>)</summary>

```
Downloading https://services.gradle.org/distributions/gradle-9.4.0-bin.zip
.............10%.............20%.............30%.............40%.............50%.............60%..............70%.............80%.............90%.............100%

Welcome to Gradle 9.4.0!

Here are the highlights of this release:
 - Java 26 support
 - Non-class-based JVM tests
 - Enhanced console progress bar

For more details see https://docs.gradle.org/9.4.0/release-notes.html

Starting a Gradle Daemon (subsequent builds will be faster)
> Task :compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Management of bidirectional association persistent attributes is deprecated and will be removed. Set the value to 'false' to get rid of this warning

> Task :processResources
> Task :classes

> Task :compileTestJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :processTestResources
> Task :testClasses
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
> Task :test

WARNING: A terminally deprecated method in sun.misc.Unsafe has been called
WARNING: sun.misc.Unsafe::objectFieldOffset has been called by net.bytebuddy.dynamic.loading.ClassInjector$UsingUnsafe$Dispatcher$CreationAction (file:/home/gradle/.gradle/caches/modules-2/files-2.1/net.bytebuddy/byte-buddy/1.17.8/af5735f63d00ca47a9375fae5c7471a36331c6ed/byte-buddy-1.17.8.jar)
WARNING: Please consider reporting this to the maintainers of class net.bytebuddy.dynamic.loading.ClassInjector$UsingUnsafe$Dispatcher$CreationAction
WARNING: sun.misc.Unsafe::objectFieldOffset will be removed in a future release
2026-03-09T14:04:26.702Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-09T14:04:26.702Z  INFO 191 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-09T14:04:26.814Z  INFO 191 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-09T14:04:26.815Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-09T14:04:26.816Z  WARN 191 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-09T14:04:26.816Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-09T14:04:26.828Z  INFO 191 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default' 
2026-03-09T14:04:26.854Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-09T14:04:26.856Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-09T14:04:26.865Z  INFO 191 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-09T14:04:26.865Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-09T14:04:26.866Z  INFO 191 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-09T14:04:26.867Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-09T14:04:26.868Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-09T14:04:26.868Z  WARN 191 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-09T14:04:26.878Z  INFO 191 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default' 
2026-03-09T14:04:26.880Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-09T14:04:26.881Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-09T14:04:26.887Z  INFO 191 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-09T14:04:26.888Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-09T14:04:26.888Z  INFO 191 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-09T14:04:26.890Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-09T14:04:26.891Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-09T14:04:26.891Z  WARN 191 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-09T14:04:26.903Z  INFO 191 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default' 
2026-03-09T14:04:26.915Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-09T14:04:26.916Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-09T14:04:26.923Z  INFO 191 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-09T14:04:26.923Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-09T14:04:26.923Z  INFO 191 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-09T14:04:26.925Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-09T14:04:26.927Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-09T14:04:26.927Z  WARN 191 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-09T14:04:26.935Z  INFO 191 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default' 
2026-03-09T14:04:26.944Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-09T14:04:26.945Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-09T14:04:26.951Z  INFO 191 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-09T14:04:26.952Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-09T14:04:26.952Z  INFO 191 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-09T14:04:26.953Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-09T14:04:26.954Z  WARN 191 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-09T14:04:26.954Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-09T14:04:26.963Z  INFO 191 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default' 
2026-03-09T14:04:26.970Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-09T14:04:26.971Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-09T14:04:26.974Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopping...
2026-03-09T14:04:26.974Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : BrokerAvailabilityEvent[available=false, SimpleBrokerMessageHandler [org.springframework.messaging.simp.broker.DefaultSubscriptionRegistry@6d67c8f9]]
2026-03-09T14:04:26.974Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopped.
2026-03-09T14:04:26.980Z  INFO 191 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-09T14:04:26.980Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-09T14:04:26.980Z  INFO 191 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-09T14:04:26.982Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-09T14:04:26.982Z  WARN 191 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-09T14:04:26.982Z  INFO 191 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-09T14:04:26.991Z  INFO 191 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default' 
2026-03-09T14:04:27.002Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-09T14:04:27.003Z  INFO 191 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.

> Task :jacocoTestReport

[Incubating] Problems report is available at: file:///booklore-api/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.4.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 7m 21s
6 actionable tasks: 6 executed
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.4.0/userguide/configuration_cache_enabling.html
```

</details>

<details>
<summary>Frontend test output (<code>ng test</code>)</summary>

```
Using Vitest configuration file: /angular-app/vitest-base.config.ts
Initial chunk files                                                            | Names                                                                       |  Raw size
spec-app-features-magic-shelf-component-magic-shelf-component.js               | spec-app-features-magic-shelf-component-magic-shelf-component               |   1.12 MB |         
polyfills.js                                                                   | polyfills                                                                   | 164.02 kB |         
spec-app-features-magic-shelf-service-book-rule-evaluator.service.js           | spec-app-features-magic-shelf-service-book-rule-evaluator.service           |  65.08 kB |         
spec-app-app.component.js                                                      | spec-app-app.component                                                      |  59.41 kB |         
chunk-QPUG245A.js                                                              | -                                                                           |  54.08 kB |         
styles.css                                                                     | styles                                                                      |  31.02 kB |         
chunk-UQBAHJ7R.js                                                              | -                                                                           |  25.09 kB |         
spec-app-features-magic-shelf-service-book-rule-evaluator-metadata-presence.js | spec-app-features-magic-shelf-service-book-rule-evaluator-metadata-presence |  23.21 kB |         
chunk-MNMYZ25K.js                                                              | -                                                                           |  15.75 kB |         
spec-app-features-magic-shelf-service-magic-shelf-utils.js                     | spec-app-features-magic-shelf-service-magic-shelf-utils                     |  11.36 kB |         
chunk-PUIDBMXW.js                                                              | -                                                                           |   6.10 kB |         
chunk-7ZEZ53JJ.js                                                              | -                                                                           |   4.92 kB |         
spec-app-core-security-auth-initializer.js                                     | spec-app-core-security-auth-initializer                                     |   4.73 kB |         
spec-app-features-book-service-library-health.service.js                       | spec-app-features-book-service-library-health.service                       |   2.95 kB |         
chunk-ZPBFDZNO.js                                                              | -                                                                           |   2.19 kB |         
chunk-ZTR6P6UG.js                                                              | -                                                                           |   2.07 kB |         
chunk-EBKNP5VF.js                                                              | -                                                                           |   1.33 kB |         
init-testbed.js                                                                | init-testbed                                                                |   1.27 kB |         
vitest-mock-patch.js                                                           | vitest-mock-patch                                                           | 704 bytes |         
spec-app-app.js                                                                | spec-app-app                                                                | 202 bytes |         

                                                                               | Initial total                                                               |   1.60 MB

Application bundle generation complete. [42.243 seconds] - 2026-03-09T14:21:26.009Z


 RUN  v4.0.18 /angular-app

 ✓  booklore  src/app/features/magic-shelf/service/magic-shelf-utils.spec.ts (46 tests) 30ms
 ✓  booklore  src/app/features/magic-shelf/service/book-rule-evaluator-metadata-presence.spec.ts (85 tests) 293ms
 ✓  booklore  src/app/features/magic-shelf/service/book-rule-evaluator.service.spec.ts (199 tests) 669ms
 ✓  booklore  src/app/app.component.spec.ts (7 tests) 458ms
 ✓  booklore  src/app/features/magic-shelf/component/magic-shelf-component.spec.ts (73 tests) 1099ms
 ✓  booklore  src/app/features/book/service/library-health.service.spec.ts (6 tests) 69ms
 ✓  booklore  src/app/core/security/auth-initializer.spec.ts (2 tests) 38ms
 ✓  booklore  src/app/app.spec.ts (1 test) 4ms

 Test Files  8 passed (8)
      Tests  419 passed (419)
   Start at  14:21:28
   Duration  91.44s (transform 7.01s, setup 55.00s, import 30.05s, tests 2.66s, environment 250.02s)

JUNIT report written to /angular-app/test-results/vitest-results.xml
```

</details>

## 📸 Screen Recording / Screenshots (MANDATORY)

> Every PR must include a **screen recording or screenshots** showing the change working end-to-end in a running local instance (both backend and frontend). This means you must have actually built, run, and tested the code yourself. PRs without visual proof will be closed without review.

image

.opus file uploaded

image

.opus file in library

image

.opus file being played in the audiobook player

---

## ✅ Pre-Submission Checklist

> **All boxes must be checked before requesting review.** Incomplete PRs will be closed without review. No exceptions.

- [X] This PR is linked to an approved issue
- [X] Code follows project style guidelines and conventions
- [X] Branch is up to date with `develop` (merge conflicts resolved)
- [X] I ran the full stack locally (backend + frontend + database) and verified the change works
- [x] Automated tests added or updated to cover changes (backend **and** frontend)
- [X] All tests pass locally and output is pasted above
- [X] Screen recording or screenshots are attached above proving the change works
- [X] PR is a single focused change (one bug fix OR one feature, not multiple unrelated changes)
- [X] PR is reasonably scoped (PRs over 1000+ changed lines will be closed, split into smaller PRs)
- [X] No unsolicited refactors, cleanups, or "improvements" are bundled in
- [X] Flyway migration versioning is correct _(if schema was modified)_
- [X] Documentation PR submitted to [booklore-docs](https://github.com/booklore-app/booklore-docs) _(if user-facing changes)_ (Docs PR [here](https://github.com/booklore-app/booklore-docs/pull/33) (published as draft)

### 🤖 AI-Assisted Contributions

> **If any part of this PR was generated or assisted by AI tools (Copilot, Claude, ChatGPT, etc.), all items below are mandatory.** You are fully responsible for every line you submit. "The AI wrote it" is not an excuse, and AI-generated PRs that clearly haven't been reviewed are the #1 reason PRs get closed.

- [X] I have read and understand every line of this PR and can explain any part of it during review
- [X] I personally ran the code and verified it works (not just trusted the AI's output)
- [X] PR is scoped to a single logical change, not a dump of everything the AI suggested
- [X] Tests validate actual behavior, not just coverage (AI-generated tests often assert nothing meaningful)
- [X] No dead code, placeholder comments, `TODO`s, or unused scaffolding left behind by AI
- [X] I did not submit refactors, style changes, or "improvements" the AI suggested beyond the scope of the issue

---

## 💬 Additional Context _(optional)_

This was my first time setting the dev environment up - smooth setup! Really liked experementing with the code.

I haven't managed to test incopatible browsers. In my understanding, .opus is supported natively by all current browsers. IE11 is not supported, and some older versions of Safari on older MacOS systems.

The Opus support relies on the RouHim `jaudiotagger` fork because the currently used upstream dependency does not provide equivalent Opus handling in Booklore's existing metadata flow.

### CODE REVIEW

File: booklore-api/src/main/java/org/booklore/service/metadata/writer/AudiobookMetadataWriter.java [hunk 1]
```diff
@@ -38,6 +38,14 @@ public class AudiobookMetadataWriter implements MetadataWriter {
...
     private final AppSettingService appSettingService;
 
+    /**
+     * Helper method to read audio files.
+     * With the RouHim jaudiotagger fork (4.0.0+), Opus files are natively supported.
+     */
+    private AudioFile readAudioFile(File audioFile) throws Exception {
+        return AudioFileIO.read(audioFile);
+    }
+
```
Comment (balazs-szucs) [inline booklore-api/src/main/java/org/booklore/service/metadata/writer/AudiobookMetadataWriter.java]: Seems unnecessary change. Your new code does same exact as the old, adds a more or less useless helper, and comment

File: booklore-api/src/main/java/org/booklore/service/reader/AudioMetadataService.java [hunk 2]
```diff
@@ -42,6 +42,14 @@ public AudiobookInfo getMetadata(BookFileEntity bookFile, Path audioPath) throws
...
     }
 
+    /**
+     * Helper method to read audio files.
+     * With the RouHim jaudiotagger fork (4.0.0+), Opus files are natively supported.
+     */
+    private AudioFile readAudioFile(Path audioPath) throws Exception {
+        return AudioFileIO.read(audioPath.toFile());
+    }
+
```
Comment (balazs-szucs) [inline booklore-api/src/main/java/org/booklore/service/reader/AudioMetadataService.java]: Same comment on this as the last.

File: booklore-api/src/main/java/org/booklore/service/reader/AudioFileUtilityService.java [hunk 3]
```diff
@@ -55,6 +55,8 @@ public String getContentType(Path audioPath) {
             return "audio/mp4";
         } else if (fileName.endsWith(".mp3")) {
             return "audio/mpeg";
+        } else if (fileName.endsWith(".opus")) {
+            return "audio/ogg";
```
Comment (balazs-szucs) [inline booklore-api/src/main/java/org/booklore/service/reader/AudioFileUtilityService.java]: ACX, we may need to look into adding support for the container MIME types not only the "strict" MIME types. E.g., for this for example we should probably handle `application/ogg` as well. This is not strictly related to this PR, just general comment.

File: booklore-api/build.gradle [hunk 4]
```diff
@@ -73,7 +74,8 @@ dependencies {
     implementation 'io.documentnode:epub4j-core:4.2.3'
 
     // --- Audio Metadata (Audiobook Support) ---
-    implementation 'net.jthink:jaudiotagger:3.0.1'
+    // Using RouHim's fork which adds Opus format support
+    implementation 'com.github.RouHim:jaudiotagger:main-SNAPSHOT'
```
Comment (balazs-szucs) [inline booklore-api/build.gradle]: https://github.com/RouHim/jaudiotagger/releases/tag/2.0.16

is the latest. This would build a build on every new commit. Which would be highly prone to breakage because 1 pushed commit is hardly considered a stable release.

File: booklore-api/src/main/java/org/booklore/service/reader/AudioMetadataService.java [hunk 5]
```diff
@@ -212,17 +223,56 @@ private AudiobookInfo buildFolderBasedAudiobookInfo(BookFileEntity bookFile, Pat
...
+        } catch (Exception e) {
+            log.debug("Could not read codec from header: {}", e.getMessage());
+        }
+
+        try {
+            channelsStr = header.getChannels();
+        } catch (Exception e) {
+            log.debug("Could not read channels from header: {}", e.getMessage());
+        }
+
```
Comment (balazs-szucs) [inline booklore-api/src/main/java/org/booklore/service/reader/AudioMetadataService.java]: This block is a DRY violation.  I think this block would be much better off using helper that extracts the header. E.g., safeExtract and then you wouldn't have to try-catch every single check because error would handled by the helper.

File: booklore-api/src/main/java/org/booklore/service/metadata/extractor/AudiobookMetadataExtractor.java [hunk 6]
```diff
@@ -173,14 +181,14 @@ private void extractAudioHeaderInfo(AudiobookMetadata.AudiobookMetadataBuilder b
             builder.durationSeconds(durationSeconds);
         }
 
-        long bitrate = header.getBitRateAsNumber();
-        if (bitrate > 0) {
-            builder.bitrate((int) bitrate);
+        Number bitrateNum = header.getBitRateAsNumber();
+        if (bitrateNum != null && bitrateNum.longValue() > 0) {
+            builder.bitrate(bitrateNum.intValue());
```
Comment (balazs-szucs) [inline booklore-api/src/main/java/org/booklore/service/metadata/extractor/AudiobookMetadataExtractor.java]: Can bitrate be null? 🤔 

Was this throwing NPEs for you, or? I don't believe auto-boxing the way to go for this, I would rather see where the actual culprit is. But this may be just a personal preference.

**Comment by balazs-szucs** [comment]
Oh... I just saw this is draft. My apologies for prematurely reviewing. 

Sorry!

File: booklore-api/build.gradle (hunk 4 above)
Comment (acx10) [inline booklore-api/build.gradle]: Nice find on RouHim's fork! One thing though, `main-SNAPSHOT` means this resolves to whatever the latest commit on `main` happens to be, so builds aren't reproducible and could break if that repo pushes a breaking change. Could you pin this to a specific commit SHA or tag instead? JitPack supports that, e.g. `com.github.RouHim:jaudiotagger:<commit-sha>`.

File: OPUS_QA_MANUAL_TESTING.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
+# Manual QA Testing Guide: Opus (.opus) Audiobook Support
```
Comment (acx10) [inline OPUS_QA_MANUAL_TESTING.md]: This is a great testing doc but probably shouldn't live in the repo itself. Could you drop this file from the PR? Happy to keep the testing steps in the PR description if you want them documented somewhere.

File: booklore-ui/src/i18n/de/reader-audiobook.json [hunk 8]
```diff
@@ -59,6 +59,7 @@
     "bookmarkExists": "Lesezeichen vorhanden",
     "bookmarkExistsDetail": "An dieser Position existiert bereits ein Lesezeichen",
     "bookmarkFailed": "Lesezeichen konnte nicht hinzugefügt werden",
-    "bookmarkDeleted": "Lesezeichen gelöscht"
+    "bookmarkDeleted": "Lesezeichen gelöscht",
+    "unsupportedOpusCodec": "Ihr Browser unterstützt Opus-Audio-Codec nicht. Bitte versuchen Sie einen anderen Browser (z. B. Chrome, Firefox, Safari 14.1+) oder konvertieren Sie die Datei in MP3."
```
Comment (acx10) [inline booklore-ui/src/i18n/de/reader-audiobook.json]: Heads up: we use Weblate for community translations, so for languages other than `en` and `es`, we only push the keys (with empty `""` values) and let translators fill them in via Weblate. Pushing translated strings directly can overwrite translator work when Weblate syncs. Could you revert the non-en/es translation files and just keep the `en` and `es` changes?

File: booklore-ui/src/app/features/readers/audiobook-player/audiobook-player.component.ts [hunk 9]
```diff
@@ -118,10 +121,21 @@ export class AudiobookPlayerComponent implements OnInit, OnDestroy {
 
     this.route.paramMap.pipe(takeUntil(this.destroy$)).subscribe(params => {
       this.bookId = +params.get('bookId')!;
+      this.checkCodecSupport();
```
Comment (acx10) [inline booklore-ui/src/app/features/readers/audiobook-player/audiobook-player.component.ts]: This runs for every audiobook, so on a browser that doesn't support opus you'd see the warning even when playing an MP3 or M4B. Might be worth checking the actual file extension of the loaded audiobook first and only showing the warning when it's an `.opus` file.

**Comment by acx10** [comment]
Also, looks like the screenshots in the PR description aren't loading (getting 404s). Could you re-upload them?

**Comment by maxbengtzen** [comment]
I've cleaned up the PR a little and resolved most of your pre-review comments. Thank you for those.

What I have is working opus support in uploader, library and metadata editing. What still feels flaky is the player support. A follow-up area worth reviewing is whether the library should expose safer defaults for Opus `AudioHeader` fields. At the moment, Booklore works around this by treating those values as optional when the library cannot provide them, so playback and metadata loading can continue even if bitrate/sample rate/channels are unavailable.

Sadly, I don't really have the time in the newst couple of weeks to investigate this further. If you want to pick this up and improve on it - feel free. If you think this is good enough as a starting point - I will certainly be able to adress any review comments.
